master/buildbot/configurators/janitor.py

```diff
@@ -40,32 +40,53 @@ def now():
 
 class LogChunksJanitor(BuildStep):
     name = 'LogChunksJanitor'
-    renderables = ["logHorizon"]
+    renderables = ["logHorizon", "buildersLogHorizon"]
 
-    def __init__(self, logHorizon):
+    def __init__(self, logHorizon=None, buildersLogHorizon=[]):
         super().__init__()
         self.logHorizon = logHorizon
+        self.buildersLogHorizon = buildersLogHorizon
 
     @defer.inlineCallbacks
     def run(self):
-        older_than_timestamp = datetime2epoch(now() - self.logHorizon)
-        deleted = yield self.master.db.logs.deleteOldLogChunks(older_than_timestamp)
-        self.descriptionDone = ["deleted", str(deleted), "logchunks"]
-        return SUCCESS
+        self.descriptionDone = []
+        if self.buildersLogHorizon != []:
+            builder_timestamps = {}
+            for builder in self.buildersLogHorizon:
+                builder_timestamps[builder] = datetime2epoch(now() - self.buildersLogHorizon[builder])
+            results = yield self.master.db.logs.deleteBuilderLogs(builder_timestamps)
+            for builder, deleted in sorted(results.items()):
+                self.descriptionDone += ["deleted", str(deleted), "logchunks from", builder, '\n']
+            self.descriptionDone.pop()
+
+        if self.logHorizon is not None:
+            older_than_timestamp = datetime2epoch(now() - self.logHorizon)
+            builders = []
+            if self.buildersLogHorizon != []:
+                builders = list(self.buildersLogHorizon.keys())
+            deleted = yield self.master.db.logs.deleteOldLogChunks(older_than_timestamp, exceptions=builders)
+            if self.descriptionDone != []:
+                self.descriptionDone += ['\n']
+            self.descriptionDone += ["deleted", str(deleted), "logchunks"]
+
+        if self.logHorizon is not None or self.buildersLogHorizon != []:
+            return SUCCESS
 
 
 class JanitorConfigurator(ConfiguratorBase):
-    def __init__(self, logHorizon=None, hour=0, **kwargs):
+    def __init__(self, logHorizon=None, hour=0, buildersLogHorizon=[], **kwargs):
```

Comment:
Shouldn't this be an empty dict? Even better, I would suggest to use `None`, because all instances of JanitorConfigurator will use the same instance of the dict and if anyone modifies it, the default value will change.

Reply:
That was a super slip! I must have gotten confused. I did thought of using None but I thought that, in some cases, having an empty dict would be better than handling None.